Cards/CardBase.gd

```diff
@@ -14,82 +14,34 @@ class_name CardBase
 ## For example, consider the card: "Deal 10 damage to all enemies, but take 3 damage"
 
 
-@export var damage_to_apply_to_target: float = 0.0
-@export var damage_to_apply_to_caster: float = 0.0
-@export var status_to_apply_to_target: Array[StatusBase]
-@export var status_to_apply_to_caster: Array[StatusBase]
-@export var affect_all_targets: bool = false
-@export var affect_all_casters: bool = false
-@export var amount_of_cards_to_draw: int = 0
-@export var amount_of_cards_to_discard: int = 0
+# @export var status_to_apply_to_target: Array[StatusBase]
+# @export var status_to_apply_to_caster: Array[StatusBase]
 @export var application_type: Enums.ApplicationType = Enums.ApplicationType.ENEMY_ONLY
 @export var card_title: String = "NULL"
 @export var card_key_art: ImageTexture = null
 @export var card_description: String = "NULL"
 
+@export var card_effects_data: Array[EffectData] = []
 
-func can_play_card(caster: Entity, target: Entity) -> bool:
-	if caster.get_party_component().can_play_on_entity(application_type, target):
-		return true
-	return false
-
-
-func on_card_play(caster: Entity, target: Entity) -> void:
-	_deal_damage(caster, target)
-	_apply_status(caster, target)
-	_draw_cards()
-	_discard_random_cards()
-	# TODO add other functionality that lots of cards may share (eg: restore AP)
-
+func _ready() -> void:
+	card_effects_data = []
 
-# override in child cards if you want to deal damage in a unique way
-func _deal_damage(caster: Entity, target: Entity) -> void:
-	# damage target
-	if damage_to_apply_to_target != 0.0:
-		_damage_entity(caster, target, damage_to_apply_to_target, affect_all_targets)
-	
-	#damage caster
-	if damage_to_apply_to_caster != 0.0:
-		_damage_entity(caster, caster, damage_to_apply_to_caster, affect_all_casters)
+# Need a parser of some sort to read a plainform card data and translate it to EffectData
+# This will call EffectData.add_effect_data() in a loop for every new effect
 
+@warning_ignore("unused_parameter")
+func parse_card_data(card_data: Dictionary) -> void:
+	# TODO
+	pass
 
-func _damage_entity(caster: Entity, target: Entity, damage_amount: float, damage_all: bool) -> void:
-	var target_damage_data: DealDamageData = DealDamageData.new()
-	target_damage_data.damage = damage_amount
-	target_damage_data.caster = caster
-	
-	# If damage_all is set, try to damage all the party members set in the party component
-	if damage_all:
-		var party: Array[Entity] = target.get_party_component().party
-		assert(party.size() > 0, "Entity has an empty party. Make sure you added party members.")
-		
-		for party_member: Entity in party:
-			party_member.get_health_component().deal_damage(target_damage_data)
-	else:
-		target.get_health_component().deal_damage(target_damage_data)
+func _apply_all_effects(entity: Entity) -> void:
+	for effect_data: EffectData in card_effects_data:
+		effect_data.apply_effect_data(entity)
```

Comment:
i'm not sure. Currently it wouldn't really make sense to put "target" there, because you can target the caster (for heal and draw for example), **BUT** later on the effects that apply to the caster would be quick-cast (meaning you wouldn't need to actually target the caster).

So TL;DR : we can put `target` even if it doesn't make sense entirely for now, because it will make sense with quick-cast

Reply:
Okay, makes sense. I'll resolve this comment.